### PR TITLE
Draft of Mail Ballot Manager API Spec

### PR DESCRIPTION
**Pull request by beausmith**

### CODE REVIEW

File: API.md [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+  packageHash: string,
+  approvedAt: datetime,
+  approvedBy: userId,
+  ballotTemplates: BallotTemplate[],
+  voters: VoterMailBallot[],
+  uploadedAt: datetime,
+}]
+```
+
+### `POST /election/new`
```
Comment (benadida) [inline API.md:94]: could just be `POST /election/`

File: API.md [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+  definition: Election,
+  packageHash: string,
+  ballotTemplates: BallotTemplate[],
+  voterMailingListCSV: string,
+}
+```
+
+## Inserts Data API
+
+### `GET /election/:electionId/inserts-data`
```
Comment (benadida) [inline API.md:107]: is this the data that configures the instructions and such that are not the ballot?

File: API.md [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+
+```
+InsertsData
+```
+
+### `POST /election/:electionId/inserts-data`
```
[the remark this answers is not in the document]
Comment (beausmith) [reply API.md:113]: Ha. I had done this then reverted.

File: API.md [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+
+```
+InsertsData
+```
+
+## Approve Printing and Mailing API
+
+### `POST /election/:electionId/approve`
```
[the remark this answers is not in the document]
Comment (beausmith) [reply API.md:121]: Yes. It was `POST /election/:electionId/approve-printing-and-mailing` but then I figured I'd make it simpler. Do you prefer the above api url?

File: API.md (hunk 1 above)
Comment (beausmith) [reply API.md:94]: 👍

File: API.md (hunk 2 above)
Comment (beausmith) [reply API.md:107]: Yes.

File: API.md [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+
+## Auth API
+
+Whatever Auth0 uses here.
+
+## Users API
+
+What user date is provided by default via Auth0?
+
+### `GET /user/`
```
[the remark this answers is not in the document]
Comment (beausmith) [reply API.md:68]: Sure.

**Comment by beausmith** [comment]
@benadida All requested changed made.

File: API.md [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+- If there is an error in the voter mailing list, how does an election official
+  fix it?
+- If an election official needs to stop the presses, can they?
+
+## Types
+
+Election as defined in `@votingworks/ballot-encoder` and the following…
+
+```
+interface VoterMailBallot {
```
Comment (eventualbuddha) [inline API.md:16]: Are any of these fields intended to be optional?

Comment (beausmith) [reply API.md:16]: Yes. Optionality should be indicated.

File: API.md [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+```
+
+## Elections API
+
+### `GET /elections/list`
+
+```
+[{
+  id: string,
+  definition: Election,
```
Comment (eventualbuddha) [inline API.md:84]: This could be rather large. Do we want to include the whole thing?

File: API.md [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+  packageHash: string,
+  printingAndMailingApprovedAt: datetime,
```
Comment (eventualbuddha) [inline API.md:86]: I'm assuming `datetime` is a ISO-8601 formatted string?

File: API.md (hunk 7 above)
Comment (beausmith) [reply API.md:84]: Yeah, was wondering what the best approach should be here. Open to suggestions.

File: API.md [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+  printingAndMailingApprovedBy: userId,
+  ballotTemplates: BallotTemplate[],
+  voters: VoterMailBallot[],
```
Comment (eventualbuddha) [inline API.md:89]: Is this all voters for an election? This is almost certainly going to be huge, right?

File: API.md (hunk 8 above)
Comment (beausmith) [reply API.md:86]: Yes, ideally.

File: API.md (hunk 9 above)
Comment (beausmith) [reply API.md:89]: Yes. Wasn't sure how to model batches, so I figured I'd just keep it simple and add 'em all to start. 

I'd like to define what exactly is huge? How many records? I've heard that React can handle a lot of data and that how much is rendered is the part that should be optimized for speed. Looking forward to learning and determining these unknowns.

File: API.md (hunk 7 above)
Comment (eventualbuddha) [reply API.md:84]: I'm guessing we'd have only fields in `/elections/list` that are required to display a list, then `/elections/<id>` would get further data? _/me wishes for graphql_

File: API.md (hunk 9 above)
Comment (eventualbuddha) [reply API.md:89]: For simplicity's sake, sure. I'm guessing on the order of 10K would be pretty comfortable. More than that and I'd start feeling anxious.

Comment (eventualbuddha) [reply API.md:89]: As for how to model batches, we'd want another endpoint like this:

```
$ curl -vv https://mail-ballot-manager.votingworks.app/election/abc123/voters
…
HTTP/1.1 200 OK
Content-Type: application/json
X-Next-Page: https://mail-ballot-manager.votingworks.app/election/abc123/voters?start=def456

[
  {
    "id": …
  },
  …
]
```

File: API.md (hunk 7 above)
Comment (beausmith) [reply API.md:84]: I had asked @benadida about using graphql, I think he said that he felt it was overkill… but I still want it. 😜

File: API.md (hunk 6 above)
Comment (beausmith) [reply API.md:16]: https://github.com/votingworks/mail-ballot-manager/pull/15

File: API.md (hunk 7 above)
Comment (benadida) [reply API.md:84]: I'm gonna keep fighting this one until I see a real use case that isn't more easily handled by simple rest :)

/elections/ is usually great for the list.

**Pull request merged**